### PR TITLE
Implement torch.special.log_ndtr

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Implements torch.special.log_ndtr 

Issue: https://github.com/pytorch/pytorch/issues/50345 

TODO:
- [x] adding proper reference to scipy implementation
- [x] double check if the changes in test/test_unary_ufuncs.py is really necessary
- [x] check setting for UnaryUfuncInfo 
cc: @kshitij12345 @mruberry

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1267,6 +1267,10 @@
   self: grad * std::sqrt(2 * M_PI) * (result.square() / 2).exp()
   result: auto_element_wise
 
+- name: special_log_ndtr(Tensor self) -> Tensor
+  self: grad / std::sqrt(2 * M_PI) * (result + self.pow(2) / 2).neg().exp()
```
Comment (SherlockNoMad) [inline tools/autograd/derivatives.yaml:1271]: this gradient definition uses 'result', thus may incur large memory usage. 
Alternatively, one can recompute with at::ndtr(self), but since ndtr is a computation heavy op, I chose to trade memory for speed.

Comment (SherlockNoMad) [reply tools/autograd/derivatives.yaml:1271]: Also, do I need to support complex number?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -14542,6 +14542,19 @@ def ref_pairwise_distance(input1, input2):
...
                    supports_fwgrad_bwgrad=True,
                    safe_casts_outputs=True),
+    UnaryUfuncInfo('special.log_ndtr',
+                   aten_name='special_log_ndtr',
+                   ref=scipy.speical.log_ndtr if TEST_SCIPY else _NOTHING,
+
+                   # TODO(bahaung): check following values
+                   decorators=(precisionOverride({torch.bfloat16: 5e-3,
+                                                  torch.float16: 5e-4}),),
+                   dtypes=all_types_and(torch.bool, torch.bfloat16),
```
Comment (SherlockNoMad) [inline torch/testing/_internal/common_methods_invocations.py:14552]: I couldn't find where the supported input dtypes are specified? It looks like the ops in pytorch doesn't have a strong-typed schema?

File: aten/src/ATen/native/Math.h [hunk 3]
```diff
@@ -2113,4 +2113,15 @@ calc_erfcx(T x)
   }
 }
 
+// TODO: add header here
+template <typename T>
+static inline C10_HOST_DEVICE T calc_log_ndtr(T x) {
```
Comment (SherlockNoMad) [inline aten/src/ATen/native/Math.h:2118]: scipy's implementation is only supporting double type, for the sake of higher precision. 
Shall I promote this to acc_type first ?

File: aten/src/ATen/native/cuda/Math.cuh [hunk 4]
```diff
@@ -276,6 +276,19 @@ const auto ndtri_string = jiterator_stringify(
   }
 ); // ndtri_string
 
+const auto log_ndtr_string = jiterator_stringify(
+  template <typename T>
+  T log_ndtr(T x) {
+    constexpr T SQRT1_2{0.707106781186547524400844362104849039};   // 1/sqrt(2)
```
Comment (SherlockNoMad) [inline aten/src/ATen/native/cuda/Math.cuh:282]: same here, promote to acc_type first?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -14542,6 +14542,19 @@ def ref_pairwise_distance(input1, input2):
                    supports_forward_ad=True,
                    supports_fwgrad_bwgrad=True,
                    safe_casts_outputs=True),
+    UnaryUfuncInfo('special.log_ndtr',
+                   aten_name='special_log_ndtr',
+                   ref=scipy.speical.log_ndtr if TEST_SCIPY else _NOTHING,
+
+                   # TODO(bahaung): check following values
+                   decorators=(precisionOverride({torch.bfloat16: 5e-3,
+                                                  torch.float16: 5e-4}),),
```
Comment (SherlockNoMad) [inline torch/testing/_internal/common_methods_invocations.py:14551]: also need to tune the precision. failing tests should report atol and rtol.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:14552]: scipy's impl supports complex input type, do we also need to support complete type?

File: test/test_unary_ufuncs.py [hunk 6]
```diff
@@ -1266,11 +1266,30 @@ def check_equal(t):
...
+        
+        # covering special implementation cases for values < -20 or > 6
+        range = (-30, 10)  
+        t = torch.linspace(*range, 1, device=device, dtype=dtype)
         check_equal(t)
 
-        # NaN, inf, -inf are tested in reference_numerics tests.
+        # Skip testing NaN, inf, -inf since they are tested in reference_numerics tests.
         info = torch.finfo(dtype)
         min, max, eps, tiny = info.min, info.max, info.eps, info.tiny
```
Comment (SherlockNoMad) [inline test/test_unary_ufuncs.py:1294]: Should these test cases be included in the OpInfo tests?

File: aten/src/ATen/native/Math.h (hunk 3 above)
Comment (mruberry) [reply aten/src/ATen/native/Math.h:2118]: PyTorch tries not to secretly use a different computation type than the input type. There are a few cases where we do promote or have distinct accumulation types, but it's rare. 

For the special operations I don't think we'd want to automatically promote to double unless the float variant of the function was consistently inaccurate, and if we did so we'd want to be sure to make the documentation very clear that the computation was always in double.

File: torch/csrc/jit/runtime/static/generated_ops.cpp [hunk 7]
```diff
@@ -2637,6 +2637,27 @@ REGISTER_OPERATOR_FUNCTOR(
       return nullptr;
     });
 
+REGISTER_OPERATOR_FUNCTOR(
```
Comment (mruberry) [inline torch/csrc/jit/runtime/static/generated_ops.cpp:2640]: Interesting -- was there a test or build failure that required this PR to add this?

File: torch/special/__init__.py [hunk 8]
```diff
@@ -547,6 +547,27 @@
     tensor([   -inf, -0.6745,  0.0000,  0.6745,     inf])
 """.format(**common_args))
 
+log_ndtr = _add_docstr(_special.special_log_ndtr,
+                   r"""
+log_ndtr(input, *, out=None) -> Tensor
+Computes the log of the area under the standard Gaussian probability density function,
```
Comment (mruberry) [inline torch/special/__init__.py:553]: Might need a space before this paragraph.

Nice short description.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 9]
```diff
@@ -504,6 +504,13 @@ static void ndtri_kernel(TensorIteratorBase& iter) {
       });
 }
 
+static void log_ndtr_kernel(TensorIteratorBase& iter) {
+  TORCH_INTERNAL_ASSERT(iter.ntensors() == 2);
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:508]: Interesting that we have all these internal asserts for the number of tensors now. I wonder where that came from?

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:14552]: The PR specifies the input datatypes using the dispatch macro and the wrapper. The dispatch macro defines the kernel for float and double, and then the wrapper specifies that boolean and integer dtypes should promote to a float type before the kernel is called. From what I'm reading I wouldn't think that bfloat16 or float16 are supported. Where are they coming from?

File: aten/src/ATen/native/UnaryOps.cpp [hunk 10]
```diff
@@ -538,8 +540,8 @@ Tensor special_sinc(const Tensor& self) { return self.sinc(); }
 namespace {
 
 inline Tensor calc_ndtr(const Tensor& self) {
-  auto x_sqrt_2 = self / std::sqrt(2.);
-  return (1 + at::erf(x_sqrt_2)) * 0.5;
+  auto x_sqrt_2 = self * M_SQRT1_2;
+  return at::erfc(-x_sqrt_2) * 0.5;
```
Comment (anjali411) [inline aten/src/ATen/native/UnaryOps.cpp:544]: is erf(-x) == -erf(x)?

File: test/test_unary_ufuncs.py (hunk 6 above)
Comment (anjali411) [reply test/test_unary_ufuncs.py:1294]: can't think of a reason not to

cc. @kshitij12345

File: test/test_unary_ufuncs.py [hunk 11]
```diff
@@ -1266,11 +1266,30 @@ def check_equal(t):
...
+        def check_equal(t):
+            # Test by comparing with scipy
+            actual = torch.special.log_ndtr(t)
+            expected = scipy.special.log_ndtr(t.cpu().numpy())
+            self.assertEqual(actual, expected)
+        
+        # covering special implementation cases for values < -20 or > 6
+        range = (-30, 10)  
+        t = torch.linspace(*range, 1, device=device, dtype=dtype)
         check_equal(t)
```
Comment (anjali411) [inline test/test_unary_ufuncs.py:1290]: what is this `check_equal` call for?

File: torch/csrc/jit/runtime/static/generated_ops.cpp [hunk 12]
```diff
@@ -2637,6 +2637,27 @@ REGISTER_OPERATOR_FUNCTOR(
...
+    aten::special_log_ndtr,
```
Comment (anjali411) [inline torch/csrc/jit/runtime/static/generated_ops.cpp:2641]: not sure why we need this

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (anjali411) [reply tools/autograd/derivatives.yaml:1271]: In case the user wants to reduce memory usage, they can use checkpointing :)

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -14542,6 +14542,19 @@ def ref_pairwise_distance(input1, input2):
                    supports_forward_ad=True,
                    supports_fwgrad_bwgrad=True,
                    safe_casts_outputs=True),
+    UnaryUfuncInfo('special.log_ndtr',
+                   aten_name='special_log_ndtr',
+                   ref=scipy.speical.log_ndtr if TEST_SCIPY else _NOTHING,
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:14547]:
```suggestion
                   ref=scipy.special.log_ndtr if TEST_SCIPY else _NOTHING,
```

File: aten/src/ATen/native/UnaryOps.cpp (hunk 10 above)
Comment (SherlockNoMad) [reply aten/src/ATen/native/UnaryOps.cpp:544]: yes... 
erf(-x) == -erf(x)
erfc(-x) == 2 - erfc(x)
erfc(x) == 1 - erf(x)

File: test/test_unary_ufuncs.py (hunk 11 above)
Comment (SherlockNoMad) [reply test/test_unary_ufuncs.py:1290]: You are right, this input range should have been covered in _float_vals.

File: torch/testing/_internal/common_methods_invocations.py (hunk 13 above)
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:14547]: yes. :)

File: test/test_unary_ufuncs.py [hunk 14]
```diff
@@ -1266,11 +1266,30 @@ def check_equal(t):
...
-        t = torch.linspace(*range, int(1e4), device=device, dtype=dtype)
+        # Skip testing NaN, inf, -inf since they are tested in reference_numerics tests.
+        info = torch.finfo(dtype)
+        min, max, eps, tiny = info.min, info.max, info.eps, info.tiny
+        t = torch.tensor([min, max, eps, tiny], dtype=dtype, device=device)
+        check_equal(t)
+    
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_special_log_ndtr_vs_scipy(self, device, dtype):
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:1280]: I think this test case is redundant with the above reference numerics tests? It's possibly those don't include min or tiny or eps or max explicitly, in which case I would add them to the extremal values test

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 9 above)
Comment (SherlockNoMad) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:508]: Indeed, I think this assert only provides marginal benefit. 
This iter instance is created by calling TensorIterator::unary_op(result, self), which is guaranteed to end up having 2 operands. 

How do we think of such defensive programming practice? Shall we remove all such tests in this file? Or swap it with TORCH_INTERNAL_ASSERT_DEBUG_ONLY? Or leave it :) ?

File: torch/csrc/jit/runtime/static/generated_ops.cpp (hunk 7 above)
Comment (SherlockNoMad) [reply torch/csrc/jit/runtime/static/generated_ops.cpp:2640]: Not needed. This should be handled by codegen.

File: torch/csrc/jit/runtime/static/generated_ops.cpp (hunk 12 above)
Comment (SherlockNoMad) [reply torch/csrc/jit/runtime/static/generated_ops.cpp:2641]: not needed. should be handled by codegen.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (SherlockNoMad) [reply tools/autograd/derivatives.yaml:1271]: I have seen a few cases where user ended up writing their own backward impl with autograd.Function. :)

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 9 above)
Comment (mruberry) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:508]: Why were they added?

Comment (SherlockNoMad) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:508]: As far as I can trace, this assert is first introduced 3 years ago... and it's a result of replacing "AT_ASSERT(ntensor == 2);" Back then, the inputs are passed in as array pointers with count: (int ntensors, char** data_). So it made sense to do such assert at that time. 

I think this line of assert is no longer needed. Just like BinaryOpKernels.cpp, these is no need to assert ntensors() == 3. 

See https://github.com/pytorch/pytorch/pull/21475/files#diff-ccaf9206c0f650871f27d32b168c31bf5706d18ee1b9c8210412cc98367ab0cdR190 
for first introduction of this line.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:14552]: indeed, these are not needed.

File: test/test_unary_ufuncs.py (hunk 14 above)
Comment (SherlockNoMad) [reply test/test_unary_ufuncs.py:1280]: I tried adding (min, max, tiny, eps) to the extremal test cases, it's causing a lot more test cases to fail... Here are some failing error messages: 

Input tensor was 3.4028234663852886e+38, torch result is **inf**, and reference result is **3.575349975927421e+38.**

For extremely tiny/large numbers, we need to redefine the meaning of values_are_close. 

I agree these cases should be added, and it should be done in a separate PR. I will raise an issue to track this task.

**Review by anjali411 — APPROVED** [review]
good to go once the CI is green

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:1272]: Avoiding recomputation is definitely the default choice we make here. So sounds good.
You don't have to support complex derivatives for such functions. If you do, you can add it to the list at https://github.com/pytorch/pytorch/blob/ffd9608963b0a6bbbed7c9fee33553628304a630/tools/autograd/gen_variable_type.py#L87 to make sure they are properly tested.

**Comment by kshitij12345** [comment]
@SherlockNoMad Thanks for the PR

It would be great if you can use `jiterator_code_stringify` to avoid duplication of kernel in both `Math.h` and `Math.cuh`.

Refer: https://github.com/pytorch/pytorch/pull/73908

File: test/test_unary_ufuncs.py (hunk 14 above)
Comment (SherlockNoMad) [reply test/test_unary_ufuncs.py:1281]: https://github.com/pytorch/pytorch/issues/74911 
issues created.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (anjali411) [reply tools/autograd/derivatives.yaml:1272]: we don't support complex numbers for functions in special module, so don't worry about it for this PR

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 9 above)
Comment (mruberry) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:508]: I agree; we could file a better engineering to follow-up and review (and likely remove) these asserts
